### PR TITLE
Slim cloudformation install flow

### PR DESCRIPTION
**Pull request by DarcyRaynerDD**
### What does this PR do?

Creates a new install flow for the forwarder using lambda layers, instead of relying on a second ZipCopier function. It makes this new install flow the default, but also keeps the ZipCopier flow as an option for regions where we can't publish lambda layers, (such as the regions in China). It also updates the installation_test script to validate both install flows.

### Motivation

The ZipCopier function, which copies an external zip from github into a local s3 bucket was proving difficult for some customers to use in restrictive AWS environments, (such as where VPC usage is mandated for all functions). This alternative flow  reduces the number of resources installed by the template by packaging code into a lambda layer.

### Testing Guidelines
You can run the ./tools/installation_test.sh script.

### Types of changes

- [ ] Bug fix
- [X] New feature
- [ ] Breaking change
- [X] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [X] This PR's description is comprehensive
- [X] This PR contains breaking changes that are documented in the description
- [X] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [X] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [X] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)
- [ ] This PR passes the unit tests 
- [ ] This PR passes the installation tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: aws/logs_monitoring/release.sh [hunk 1]
```diff
@@ -98,6 +117,11 @@ if [ "$ACCOUNT" = "prod" ] ; then
     echo "Releasing aws-dd-forwarder-${VERSION} to GitHub..."
     go get github.com/github/hub
     hub release create -a $BUNDLE_PATH -m "aws-dd-forwarder-${VERSION}" aws-dd-forwarder-${VERSION}
+    
+    # Upload the sandbox layers
```
Comment (nhinsch) [inline aws/logs_monitoring/release.sh]: I think this comment is wrong -- should be prod

File: aws/logs_monitoring/tools/build_bundle.sh [hunk 2]
```diff
@@ -34,12 +34,13 @@ function make_path_absolute {
 
 function docker_build_zip {
     # Args: [python version] [zip destination]
-
     destination=$(make_path_absolute $2)
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/build_bundle.sh]: can we change `destination` to `zip_destination`?

File: aws/logs_monitoring/tools/build_bundle.sh [hunk 3]
```diff
@@ -51,13 +52,25 @@ function docker_build_zip {
...
 
     rm -rf $temp_dir
-    echo "Done creating archive $destination"
+    echo "Done creating forwarder zip archive $destination"
+
+    temp_dir=$(mktemp -d)
+    SUB_DIRECTORY=python
+    mkdir $temp_dir/$SUB_DIRECTORY
+
+    # Run the image by runtime tag, tar its generatd `python` directory to sdout,
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/build_bundle.sh]: Nit: generatd -> generated

File: aws/logs_monitoring/tools/list_layers.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    if [[ ! "$AVAILABLE_REGIONS" == *"$2"* ]]; then
+        >&2 echo "Could not find $2 in available regions:" $AVAILABLE_REGIONS
+        >&2 echo ""
+        >&2 echo "EXITING SCRIPT."
+        exit 1
+    fi
+    REGIONS=($2)
+fi
+
+# Check region arg
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/list_layers.sh:33]: This comment talks about a "region" arg, but I think the code refers to the first arg ("layer-name"). However, there is only one possible layer name so maybe that argument and all of this code can be removed?

File: aws/logs_monitoring/tools/publish_layers.sh [hunk 5]
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
+
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019 Datadog, Inc.
+
+# Publish the datadog python lambda layer across regions, using the AWS CLI
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/publish_layers.sh]: Looks like this was copypasted from the Python layer, should be updated for the Forwarder

File: aws/logs_monitoring/tools/publish_layers.sh [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    echo "Region specified: $REGIONS"
+    if [[ ! "$AVAILABLE_REGIONS" == *"$REGIONS"* ]]; then
+        echo "Could not find $REGIONS in available regions: $AVAILABLE_REGIONS"
+        echo ""
+        echo "EXITING SCRIPT."
+        exit 1
+    fi
+fi
+
+# Determine the target layers
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/publish_layers.sh:44]: Likewise only one possible layer here

File: .github/workflows/lint.yml [hunk 7]
```diff
@@ -27,7 +27,10 @@ jobs:
           pip install black
           black --check ./aws/logs_monitoring
 
-      - name: Lint CloudFormation template with cfn-lint
-        uses: scottbrenner/cfn-lint-action@master
-        with:
-          args: "aws/logs_monitoring/template.yaml"
+      - name: Setup Cloud Formation Linter with Latest Version
```
Comment (nhinsch) [inline .github/workflows/lint.yml:30]: Looks like this doesn't necessarily use the latest version (it will always use version 2). I would just stick to "Setup Cloud Formation Linter"

File: aws/logs_monitoring/README.md [hunk 8]
```diff
@@ -365,6 +365,10 @@ To test different patterns against your logs, turn on [debug logs](#troubleshoot
 
 `AdditionalTargetLambdaARNs`
 : Comma separated list of Lambda ARNs that will get called asynchronously with the same `event` the Datadog Forwarder receives.
+`InstallFromLayer`
+: Disable to use legacy installation flow, (via Github), for regions where Lambda layers aren't supported, such as regions in China. Defaults to true.
```
Comment (nhinsch) [inline aws/logs_monitoring/README.md]: I would reword this, I don't think it's obvious what `legacy installation flow, (via Github)` entails.

Comment (nhinsch) [reply aws/logs_monitoring/README.md]: Also, I think "regions where Lambda layers aren't supported, such as regions in China" is a bit misleading. It looks like Lambda layers are supported in China, but we just don't publish our layers there.

File: aws/logs_monitoring/README.md [hunk 9]
```diff
@@ -365,6 +365,10 @@ To test different patterns against your logs, turn on [debug logs](#troubleshoot
 
 `AdditionalTargetLambdaARNs`
 : Comma separated list of Lambda ARNs that will get called asynchronously with the same `event` the Datadog Forwarder receives.
+`InstallFromLayer`
```
Comment (nhinsch) [inline aws/logs_monitoring/README.md]: I would rename this to `InstallAsLayer`. `InstallFromLayer` makes it sound like we are somehow copying something off of a layer.

File: aws/logs_monitoring/template.yaml [hunk 10]
```diff
@@ -52,6 +53,17 @@ Parameters:
     Type: String
     Default: ""
     Description: DO NOT CHANGE unless you know what you are doing. Override the default location of the function source code.
+  InstallFromLayer:
```
Comment (nhinsch) [inline aws/logs_monitoring/template.yaml]: I left some comments about these parameters on the README file, but those comments apply to the template as well.

File: aws/logs_monitoring/README.md [hunk 11]
```diff
@@ -365,6 +365,10 @@ To test different patterns against your logs, turn on [debug logs](#troubleshoot
...
+: Disable to use legacy installation flow, (via Github), for regions where Lambda layers aren't supported, such as regions in China. Defaults to true.
+`LayerARN`
+: ARN for the layer containing the forwarder code. If empty, the script will use the version of the layer the forwarder was published with. Defaults to empty.
```
Comment (nhinsch) [inline aws/logs_monitoring/README.md:372]: These docs make it sound like the `InstallFromLayer` option determines whether or not we install from the layer. But looking at the template, it seems that we automatically don't install from the layer when we detect the installation is happening in AWS China or the `SourceZipUrl` parameter is set.

Also, why do we assume that the customer wants to use a Zip if they set a `SourceZipUrl`, but we don't necessarily assume they want to use a layer if they set `LayerARN`?

File: aws/logs_monitoring/README.md (hunk 9 above)
Comment (nhinsch) [reply aws/logs_monitoring/README.md]: We could also flip it and make this `InstallAsZip`? With a default of `false`?

File: aws/logs_monitoring/tools/list_layers.sh [hunk 12]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/list_layers.sh:1]: Are you sure that the `list_layers.sh` and `add_new_region.sh` scripts are needed? It looks to me like the `publish_layers.sh` script uses all available regions.

File: aws/logs_monitoring/tools/list_layers.sh (hunk 4 above)
Comment (DarcyRaynerDD) [reply aws/logs_monitoring/tools/list_layers.sh:33]: This file is pretty much copy pasted from the python layer. I think it will make it easier to port changes back if we avoid unnecessary changes. Wondering if it might make sense to abstract our layer tooling into a reusable utility that we share between codebases, and only keep settings like LAYER_NAME in the local repo.

File: .github/workflows/lint.yml (hunk 7 above)
Comment (DarcyRaynerDD) [reply .github/workflows/lint.yml:30]: It's v2 of the github action, but I think it does use the latest version of the linter, at least based on the actions documentation. https://github.com/ScottBrenner/cfn-lint-action#upgrading-from-version-1

File: aws/logs_monitoring/tools/publish_layers.sh [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019 Datadog, Inc.
+
+# Publish the datadog python lambda layer across regions, using the AWS CLI
+# Usage: VERSION=5 REGIONS=us-east-1 LAYERS=Datadog-Python27 publish_layers.sh
+# VERSION is required.
+set -e
+
```
Comment (nhinsch) [inline aws/logs_monitoring/tools/publish_layers.sh:12]: It looks like this script requires `VERSION` and `FORWARDER_VERSION`. The Forwarder version seems like a semantic version number whereas version is the layer version. However, the usage of these two terms isn't consistent across all of the scripts.

Could we switch to using `LAYER_VERSION` and `FORWARDER_VERSION` everywhere, so it is easy to keep straight which is which in all of the scripts?

File: aws/logs_monitoring/tools/list_layers.sh (hunk 4 above)
Comment (nhinsch) [reply aws/logs_monitoring/tools/list_layers.sh:33]: Yeah, a reusable layer deploy tool could be a smart idea.

In the mean time, I buy that it's reasonable to leave this extra argument here (I don't think it does any harm). Some of my other comments still apply, I think there's potential for confusion with some of the copy-pasted stuff.

File: aws/logs_monitoring/tools/list_layers.sh (hunk 12 above)
Comment (DarcyRaynerDD) [reply aws/logs_monitoring/tools/list_layers.sh:1]: list_layers is useful for local development, even if it isn't strictly needed during the publish flow. add_new_region is useful because it grabs a current layer zip and redeploys that, without having to go back and rebuild from the last release commit.

File: aws/logs_monitoring/README.md (hunk 11 above)
Comment (DarcyRaynerDD) [reply aws/logs_monitoring/README.md:372]: I could add a check the the China condition, to see if LayerARN was manually set. The reason for having the China condition there is so user's won't have to think decide which flow to use by default. I can probably remove the SourceZipUrl condition.

File: aws/logs_monitoring/README.md [hunk 14]
```diff
@@ -365,6 +365,10 @@ To test different patterns against your logs, turn on [debug logs](#troubleshoot
 
 `AdditionalTargetLambdaARNs`
 : Comma separated list of Lambda ARNs that will get called asynchronously with the same `event` the Datadog Forwarder receives.
+`InstallAsLayer`
+: Whether to use the layer based installation flow. Set to false to use our legacy installation flow, which installs a second function that copies the forwarder code from Github to an S3 bucket. Disable this option if installing into AWS China regions, where we don't publish layers. Defaults to true.
```
Comment (nhinsch) [inline aws/logs_monitoring/README.md]: Nits:
layer based -> layer-based
Github -> GitHub

Also, this says "Disable this option if installing into AWS China regions, where we don't publish layers." However, looking at the template, it looks like you don't actually need to disable this option when publishing to China -- it happens automatically.

File: aws/logs_monitoring/release.sh [hunk 15]
```diff
@@ -63,7 +81,7 @@ if [ "$ACCOUNT" = "prod" ] ; then
 
     # Confirm to proceed
     echo
-    read -p "About to bump the version from ${CURRENT_VERSION} to ${VERSION}, create a release aws-dd-forwarder-${VERSION} on GitHub and upload the template.yaml to s3://${BUCKET}/aws/forwarder/${VERSION}.yaml. Continue (y/n)?" CONT
+    read -p "About to bump the version from ${CURRENT_VERSION} to ${FORWARDER_VERSION}, create a release aws-dd-forwarder-${FORWARDER_VERSION} on GitHub and upload the template.yaml to s3://${BUCKET}/aws/forwarder/${FORWARDER_VERSION}.yaml. Continue (y/n)?" CONT
```
Comment (nhinsch) [inline aws/logs_monitoring/release.sh]: Should we update this message to also say that we are publishing a new layer?

**Review by nhinsch — APPROVED** [review]
Looks great! Just a couple minor nits

File: aws/logs_monitoring/tools/publish_layers.sh (hunk 5 above)
Comment (nhinsch) [reply aws/logs_monitoring/tools/publish_layers.sh]: This comment should still be updated
